### PR TITLE
Update Result docs to the new guarantees

### PR DESCRIPTION
**Pull request by Lokathor**
**Note:** This PR did get merged, even though Github claims it did not.

The `Option` docs already explain the guarantees given in [RFC 3391](https://github.com/rust-lang/rfcs/blob/master/text/3391-result_ffi_guarantees.md), so all that we need is a paragraph saying that some `Result` type combinations will also qualify.

Part of https://github.com/rust-lang/rust/issues/110503

### CODE REVIEW

File: library/core/src/result.rs [hunk 1]
```diff
@@ -224,6 +224,21 @@
 //! [`Err(E)`]: Err
 //! [io::Error]: ../../std/io/struct.Error.html "io::Error"
 //!
+//! # Representation
+//!
+//! In some cases, [`Result<T, E>`] will gain the same size, alignment, and ABI
+//! guarantees as [`Option<T>`] has. One of either the `T` or `E` type must be a
```
Comment (ChrisJefferson) [inline library/core/src/result.rs:230]: I found this a bit confusing the first couple of times I read it, because I thought the `T` in `Result<T,E>` and in `Option<T>` were the same `T`.

Maybe change to `Option<U>` (or other, then explain the non-empty one is `U`?

Comment (Lokathor) [reply library/core/src/result.rs:230]: that's a good point but i just turned off the PC so I'll do that tomorrow

File: library/core/src/result.rs [hunk 2]
```diff
@@ -1,9 +1,9 @@
 //! Error handling with the `Result` type.
 //!
 //! [`Result<T, E>`][`Result`] is the type used for returning and propagating
-//! errors. It is an enum with the variants, [`Ok(T)`], representing
-//! success and containing a value, and [`Err(E)`], representing error
-//! and containing an error value.
+//! errors. It is an enum with the variants, [`Ok(T)`], representing success and
```
Comment (thomcc) [inline library/core/src/result.rs:4]: This change is unrelated. You've changed the formatting all over the place in this patch, can you make it so that the only change is the new addition?

**Review by thomcc — COMMENTED** [review]
Looks fine except for the unrelated changes.

File: library/core/src/result.rs (hunk 2 above)
Comment (tgross35) [reply library/core/src/result.rs:4]: The formatting change is already split to a separate commit from the new content, doesn't seem like those changes justify their own PR (separation of concerns vs. review resource batching...)

**Comment by the8472** [comment]
The option module page currently is the central place for NPO documentation. If Result gets those too then things become more scattered. Maybe the Option page should also link to the Result page.

File: library/core/src/result.rs (hunk 2 above)
Comment (Lokathor) [reply library/core/src/result.rs:4]: I just used my editor to rewrap the comments while editing the new paragraph, and it also rewrapped these other paragraphs. I didn't think it mattered since the generated rustdoc is the same, but I can adjust it if people think it matters.

**Comment by Lokathor** [comment]
How exactly do you think that would look @the8472 ?

The current wording is intended to be a "one way" lookup. "if X is the case, you can count on Y too", but since the guaranteed effects are part of Option then I'm not sure how Option would refer to Result.

**Comment by the8472** [comment]
Maybe

> Under some conditions the above types `T`  are also null pointer optimized when wrapped in a [Result#Representation]

**Review by dtolnay — APPROVED** [review]
LGTM. Thank you!

**Comment by RalfJung** [comment]
Why is this "closed", not "merged"? Something seems to have gone wrong?

(Also the commits would ideally have been squashed before landing this, no need to have all these 1-line-fixup commits perpetually in the git history...)

**Comment by Lokathor** [comment]
i thought it was handled by bors >.>

**Comment by RalfJung** [comment]
Normally bors does nothing, Github marks a PR as merged when its commits all landed.

**Comment by Lokathor** [comment]
https://github.com/rust-lang/rust/blob/master/library/core/src/result.rs#L231-L250

the PR's changes do seem to be in the `master` branch

**Comment by RalfJung** [comment]
Yeah, 9b480da36790ea8314b77cf97ecaf7653afb3a1a got merged. No idea why this is not reflected in the PR state.
